synapse/push/push_tools.py

```diff
@@ -49,7 +50,33 @@ async def get_badge_count(store: DataStore, user_id: str, group_by_room: bool) -
 async def get_context_for_event(
     storage: StorageControllers, ev: EventBase, user_id: str
 ) -> Dict[str, str]:
-    ctx = {}
+    ctx: Dict[str, str] = {}
+
+    if ev.internal_metadata.outlier:
+        # We don't have state for outliers, so we can't compute the context
+        # except for invites with `invite_room_state` set..
+        if (
+            ev.type != EventTypes.Member
+            or ev.content.get("membership") != Membership.INVITE
+        ):
+            return ctx
+
+        invite_room_state = ev.unsigned.get("invite_room_state", [])
+
+        # Ideally we'd reuse the logic in `calculate_room_name`, but that gets
+        # complicated to handle partial events vs pulling events from the DB.
+        for state_dict in invite_room_state:
+            type_tuple = (state_dict["type"], state_dict.get("state_key"))
+            if type_tuple == (EventTypes.Member, ev.sender):
+                display_name = state_dict["content"].get("displayname")
+                if display_name:
+                    ctx["sender_display_name"] = display_name
+            elif type_tuple == (EventTypes.Name, ""):
+                room_name = state_dict["content"].get("name")
+                if room_name:
+                    ctx["name"] = room_name
+
+        return ctx
 
     room_state_ids = await storage.state.get_state_ids_for_event(ev.event_id)
 
```

Comment:
The `ctx["sender_display_name"] = name_from_member_event(sender_state_event)` below seems flawed as the `sender` isn't necessarily the person who sent the event.

Reply:
Ahh, nvm. We grab the relevant `sender` member event. I was thinking it was pulling from the given `event`